### PR TITLE
[Merged by Bors] - feat: port GroupTheory.Subgroup.Basic

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Vierkantor** [comment]
It's somewhat surprising that this seems to have more problems than `LinearMap` given that I expected unbundled inheritance as in `MulMemClass` etc. to be easier on Lean 4's `outParam` support than bundled as in `LinearMapClass`. I'll take a look, see if I can fix some of the issues.

**Comment by Vierkantor** [comment]
I think I see the issue: `MulMemClass` etc. should not have an `outParam` and instead leave figuring out the type of elements to `SetLike`. Let's see if that helps...

**Comment by Vierkantor** [comment]
I spun out the `outParam` fix to a new PR: #1832. Next, I'll try and see if we can backport this fix to Lean 3, for easier porting in the future.

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,3736 @@
...
+
+@[to_additive]
+theorem map_symm_eq_iff_map_eq {H : Subgroup N} {e : G ≃* N} : H.map ↑e.symm = K ↔ K.map ↑e = H :=
+  by
+  constructor <;> rintro rfl
+  ·
+    rw [map_map, ← MulEquiv.coe_monoidHom_trans, MulEquiv.symm_trans_self,
+      MulEquiv.coe_monoidHom_refl, map_id]
+  ·
+    rw [map_map, ← MulEquiv.coe_monoidHom_trans, MulEquiv.self_trans_symm,
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:1510]:
```suggestion
  · rw [map_map, ← MulEquiv.coe_monoidHom_trans, MulEquiv.symm_trans_self,
      MulEquiv.coe_monoidHom_refl, map_id]
  · rw [map_map, ← MulEquiv.coe_monoidHom_trans, MulEquiv.self_trans_symm,
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,3736 @@
...
+
+/-- For any subgroups `H` and `K`, view `H ⊓ K` as a subgroup of `K`. -/
+@[to_additive "For any subgroups `H` and `K`, view `H ⊓ K` as a subgroup of `K`."]
+def subgroupOf (H K : Subgroup G) : Subgroup K :=
+  H.comap K.subtype
+#align subgroup.subgroup_of Subgroup.subgroupOf
+#align add_subgroup.add_subgroup_of AddSubgroup.addSubgroupOf
+
+/-- If `H ≤ K`, then `H` as a subgroup of `K` is isomorphic to `H`. -/
+@[to_additive "If `H ≤ K`, then `H` as a subgroup of `K` is isomorphic to `H`.", simps]
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:1611]: Does this transfer the `simps` as well, or does it not need to?

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,3736 @@
...
+@[to_additive prod
+      "Given `AddSubgroup`s `H`, `K` of `AddGroup`s `A`, `B` respectively, `H × K`
+      as an `AddSubgroup` of `A × B`."]
+def prod (H : Subgroup G) (K : Subgroup N) : Subgroup (G × N) :=
+  { Submonoid.prod H.toSubmonoid K.toSubmonoid with
+    inv_mem' := fun hx => ⟨H.inv_mem' hx.1, K.inv_mem' hx.2⟩ }
+#align subgroup.prod Subgroup.prod
+#align add_subgroup.prod AddSubgroup.prod
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:1724]: ```suggestion```

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,3736 @@
...
+    normalClosure ↑(closure s) = normalClosure s :=
+  le_antisymm (normalClosure_le_normal closure_le_normalClosure) (normalClosure_mono subset_closure)
+#align
+  subgroup.normal_closure_closure_eq_normal_closure
+  Subgroup.normalClosure_closure_eq_normalClosure
+
+/-- The normal core of a subgroup `H` is the largest normal subgroup of `G` contained in `H`,
+as shown by `Subgroup.normalCore_eq_supᵢ`. -/
+def normalCore (H : Subgroup G) : Subgroup G
+    where
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:2550]:
```suggestion
def normalCore (H : Subgroup G) : Subgroup G where
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,3736 @@
...
+    simp only [comp_apply, liftOfRightInverseAux_comp_apply, Subtype.coe_mk]
+  right_inv φ := by
+    ext b
+    simp [liftOfRightInverseAux, hf b]
+#align monoid_hom.lift_of_right_inverse MonoidHom.liftOfRightInverse
+#align add_monoid_hom.lift_of_right_inverse AddMonoidHom.liftOfRightInverse
+
+/-- A non-computable version of `MonoidHom.liftOfRightInverse` for when no computable right
+inverse is available, that uses `Function.surjInv`. -/
+@[simp,
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3374]: Does this need `to_additive`-izing?

File: Mathlib/GroupTheory/Subgroup/Basic.lean (hunk 2 above)
Comment (fpvandoorn) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:1604]: no, this needs fixing

File: Mathlib/GroupTheory/Subgroup/Basic.lean (hunk 5 above)
Comment (fpvandoorn) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3365]: this should raise a linter error... Yes, it does.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**